auth/cache.go

```diff
@@ -0,0 +1,162 @@
+package auth
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"os"
+	"path/filepath"
+	"time"
+
+	"golang.org/x/oauth2"
+)
+
+const (
+	MaxAge = time.Hour * 24
+
+	defaultDirectory     = ".config/hcp"
+	testDirectory        = "hcptest"
+	fileName             = "credentials.json"
+	directoryPermissions = 0755
+
+	envVarCacheTestMode = "HCP_CACHE_TEST_MODE"
+)
+
+type Cache struct {
+	// AccessToken is the bearer token used to authenticate to HCP.
+	AccessToken string `json:"access_token,omitempty"`
+
+	// RefreshToken is used to get a new access token.
+	RefreshToken string `json:"refresh_token,omitempty"`
+
+	// Expiry is when the access token will expire.
+	Expiry time.Time `json:"expiry,omitempty"`
+
+	// MaxAge is the session limit.
+	MaxAge time.Duration `json:"max_age,omitempty"`
+}
+
+// Write saves HCP auth data in a common location in the home directory.
+func Write(token oauth2.Token) error {
+	credentialPath, credentialDirectory, err := getCredentialPaths()
+	if err != nil {
+		return fmt.Errorf("failed to retrieve credential path and directory: %v", err)
+	}
+
+	err = os.MkdirAll(credentialDirectory, directoryPermissions)
+	if err != nil {
+		return fmt.Errorf("failed to create credential directory: %v", err)
+	}
+
+	// Write access token, refresh_token, expiry, max age to credentials file.
+	credentials := &Cache{
+		AccessToken:  token.AccessToken,
+		RefreshToken: token.RefreshToken,
+		Expiry:       token.Expiry,
+		MaxAge:       MaxAge,
+	}
+
+	credentialsJSON, err := json.Marshal(credentials)
+	if err != nil {
+		return fmt.Errorf("failed to marshal the struct to json: %v", err)
+	}
+
+	err = os.WriteFile(credentialPath, credentialsJSON, directoryPermissions)
+	if err != nil {
+		return fmt.Errorf("failed to write credentials to the cache file: %v", err)
+	}
+
+	return nil
+}
+
+// Read opens the saved HCP auth data and returns the token.
+func Read() (*Cache, error) {
+	credentialPath, _, err := getCredentialPaths()
+	if err != nil {
+		return nil, fmt.Errorf("failed to retrieve credential path and directory: %v", err)
+	}
+
+	rawJSON, err := os.ReadFile(credentialPath)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read file from user's credential path: %v", err)
+	}
+
+	cacheFromJSON, err := jsonToToken(rawJSON)
+	if err != nil {
+		return nil, fmt.Errorf("bad format: %v", err)
+	}
+
+	return cacheFromJSON, nil
```

Comment:
I think we might want another function to return the token data as an oauth2.Token, since we can pass that directly into an oauth config struct 🤔 (Like how we pass the token in the second parameter [here](https://github.com/hashicorp/hcp-sdk-go/blob/3bf7fd733b82d13780d41e4cf7cd94ff0d3fb4b5/config/new.go#L122))

Reply:
In that case does it make more sense to just return the `oauth2.token` from the `Read()` function since the `Write()` function takes that in as a param?

Reply:
The only problem there is that we do still need to return the MaxAge (aka SessionExpiry), which can't be represented in the oauth2 Token. Maybe compromise is a ReadToken and ReadSessionExpiry? 💭